### DISCUSSION TITLE
seccomp notifier metric cardinality

### DISCUSSION THREAD

**Discussion by pdf** [cri-o/cri-o#6418] [General]
Looking at the recently added seccomp metrics, it looks like the `syscalls` label has essentially unbounded cardinality, since it appears that syscalls and counts are concatenated as a string value, e.g.:

```
container_runtime_crio_containers_seccomp_notifier_count_total{name="…",syscalls="chroot (1x)"} 1
container_runtime_crio_containers_seccomp_notifier_count_total{name="…",syscalls="chroot (2x), swapoff (2x)"} 1
container_runtime_crio_containers_seccomp_notifier_count_total{name="…",syscalls="chroot (1x), swapoff (2x)"} 1
container_runtime_crio_containers_seccomp_notifier_count_total{name="...",syscalls="swapoff (5x)"} 1
```

So, every possible combination of syscall and count may published to this label, making the possible value permutations infinite.

I'm not sure what the ideal structure for these metrics should be, but I don't think the current metrics should stand as they are.

**Comment by saschagrunert** [comment]
Yeah, some features are _unlikely_ to produce many metrics and in this case we decided to add them as they are. We could switch them to a different format, like:

```
container_runtime_crio_containers_seccomp_notifier_count_total{name="abc",syscalls="chroot"} 2
container_runtime_crio_containers_seccomp_notifier_count_total{name="abc",syscalls="swapoff"} 1
```

The question is would that make anything easier to debug… :thinking:

**Comment by pdf** [comment]
I don't think it would make things _harder_ to debug, which is probably the more important consideration.

The real killer for cardinality is the counts - with just concatenated (sorted) syscall names, the label values are at least bounded by the permutations of syscall names, though I still don't love that.

If you stick to a single label per point it also opens up the possibility to query by syscall without having to resort to regexp, however useful that might or might not be.

**Reply by haircommander** [reply]
WDYT @saschagrunert ? @pdf do you have any interest in updating the metric?

**Reply by saschagrunert** [reply]
Sounds good to me! :+1:

**Reply by pdf** [reply]
@haircommander I can update if you'd like me to. The label name should probably be singular to reflect the new values, ie:

```diff
--- a/labels    2022-12-11 11:00:55.848154596 +1100
+++ b/labels    2022-12-11 11:01:13.000192573 +1100
@@ -1 +1 @@
-container_runtime_crio_containers_seccomp_notifier_count_total{name="abc",syscalls="chroot"} 2
+container_runtime_crio_containers_seccomp_notifier_count_total{name="abc",syscall="chroot"} 2
```

Any other thoughts before I start? I'm rather time-poor atm, but should be able to squeeze it in this week sometime.

**Reply by haircommander** [reply]
SGTM thanks for taking this on!

**Reply by haircommander** [reply]
@pdf did you still want to take this on? we'll be looking to cut 1.26 soon and I'd really like to get this in

**Reply by pdf** [reply]
@haircommander I had a look last night and the change itself looks straight-forward, but I had trouble getting reliable test runs locally so I think I'd have to build a dedicated env to get the test suite sorted, likely some days out before I can allocate the time.

**Reply by haircommander** [reply]
FYI, I'm looking to cut 1.26.0 early next week, and I'd love to have this fixed by then. If you need, you could use our CI as a way to check against the tests. not the best practice usually, but good in a pinch like this. WDYT?

**Reply by pdf** [reply]
I should have some time over the weekend, will try to sort it locally, otherwise I'll take the push it and see approach. Totally understand the desire to get it sorted before the current metric makes release.

**Reply by haircommander** [reply]
Thank you for taking this on and conforming to the now tight schedule :smile:

**Reply by pdf** [reply]
@haircommander I couldn't work out the test suite in time, drafted a PR for testing on CI at #6456
